hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/keyvalue/TestKeyValueHandler.java

```diff
@@ -499,10 +578,121 @@ public void testReconcileContainer(ContainerLayoutVersion layoutVersion) throws
+    DatanodeDetails peer1 = MockDatanodeDetails.randomDatanodeDetails();
+    DatanodeDetails peer2 = MockDatanodeDetails.randomDatanodeDetails();
+    DatanodeDetails peer3 = MockDatanodeDetails.randomDatanodeDetails();
+    when(mockDnClient.getContainerChecksumInfo(anyLong(), any())).thenReturn(null);
+    keyValueHandler.reconcileContainer(mockDnClient, container, Sets.newHashSet(peer1, peer2, peer3));
+    // Make sure all the replicas are used for reconciliation.
+    Mockito.verify(mockDnClient, times(3)).getContainerChecksumInfo(anyLong(), any());
     Assertions.assertEquals(1, icrCount.get());
   }
 
+  @ParameterizedTest
+  @MethodSource("corruptionValues")
+  public void testFullContainerReconciliation(int numBlocks, int numChunks) throws Exception {
+    KeyValueHandler kvHandler = createKeyValueHandler(testRoot);
+    ContainerChecksumTreeManager checksumManager = kvHandler.getChecksumManager();
+    DNContainerOperationClient dnClient = mock(DNContainerOperationClient.class);
+    XceiverClientManager xceiverClientManager = mock(XceiverClientManager.class);
+    TokenHelper tokenHelper = new TokenHelper(new SecurityConfig(conf), null);
+    when(dnClient.getTokenHelper()).thenReturn(tokenHelper);
+    when(dnClient.getXceiverClientManager()).thenReturn(xceiverClientManager);
+    final long containerID = 100L;
+    // Create 3 containers with 10 blocks each and 3 replicas.
+    List<KeyValueContainer> containers = createContainerWithBlocks(kvHandler, containerID, 15, 3);
+    assertEquals(3, containers.size());
+
+    // Introduce corruption in each container on different replicas.
+    introduceCorruption(kvHandler, containers.get(1), numBlocks, numChunks, false);
+    introduceCorruption(kvHandler, containers.get(2), numBlocks, numChunks, true);
+
+    // Without reconciliation, checksums should be different because of the corruption.
+    Set<Long> checksumsBeforeReconciliation = new HashSet<>();
+    for (KeyValueContainer kvContainer : containers) {
+      kvHandler.createContainerMerkleTree(kvContainer);
+      Optional<ContainerProtos.ContainerChecksumInfo> containerChecksumInfo =
+          checksumManager.read(kvContainer.getContainerData());
+      assertTrue(containerChecksumInfo.isPresent());
+      checksumsBeforeReconciliation.add(containerChecksumInfo.get().getContainerMerkleTree().getDataChecksum());
+    }
+    // There should be more than 1 checksum because of the corruption.
+    assertTrue(checksumsBeforeReconciliation.size() > 1);
+
+    List<DatanodeDetails> datanodes = Lists.newArrayList(randomDatanodeDetails(), randomDatanodeDetails(),
+        randomDatanodeDetails());
+
+    // Setup mock for each datanode network calls needed for reconciliation.
+    try (MockedStatic<ContainerProtocolCalls> containerProtocolMock = Mockito.mockStatic(ContainerProtocolCalls.class);
+         MockedStatic<DNContainerOperationClient> dnClientMock = Mockito.mockStatic(DNContainerOperationClient.class)) {
```

Comment:
We should be able to mock only `ContainerProtocolCalls`. The peer node information can be retrieved from the `pipeline` field in the xceiver client given to each static method.